### PR TITLE
Blockinfile multiline search

### PR DESCRIPTION
**Pull request by OscarBell**
##### SUMMARY
Add option `multiline_search`, which enables the ability to use a multiline search regex for finding the correct spot in the file where the insert should happen. Instead of searching per line, when this option is enabled, the search will be executed within the whole file.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
blockinfile

##### ADDITIONAL INFORMATION
Sometimes you might want to insert text before or after lines of text which can only be found via a multiline search. This feature request adds an option `multiline_search` which, when enabled, uses re.MULTILINE to search in the whole file instead of searching per line in the file.

If the `multiline_search` parameter is not specified or has the value False|no the behaviour of the module is as before. Only when enabled the search mechanism will be different.

### CODE REVIEW

**Comment by jborean93** [comment]
Is it possible to do this today by prefixing the regex with the mulitline line `(^)` so that it does search in multiple lines?

**Comment by jborean93** [comment]
Is it possible to search the requested regex string to determine if the multiline flag is present `(^)`. This saves us from having to add a new option.

**Comment by OscarBell** [comment]
@jborean93 Good idea, I've tested this with searching the string for multiline flags ^ or $ and that seems to work. Will update the code

**Comment by OscarBell** [comment]
Need to solve the issues with the failing test cases because the way we detect multiline search has changed.

**Comment by jborean93** [comment]
The multiline regex flag I was talking about is if `(?m)` was specified at the start of the pattern. I was thinking if the pattern did start with this then we would do the search on the full string rather than line by line. Essentially `re.search('(?m)pattern', string)` is the same as `re.search('pattern', string, flags=re.M)`.

The PR also needs a changelog fragment to document this new feature in the next release https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment.

**Comment by OscarBell** [comment]
@jborean93: 

- Ok, no problem, I can change the code to make `(?m)` as the _flag_ for the user to define multiline search or not. Is that what you want? In that case I'll update the code for this solution and update the documentation in the module
- The changeling fragment I was aware of but I was not sure when to create it. In the next push I'll add this to the PR

**Comment by jborean93** [comment]
> Ok, no problem, I can change the code to make (?m) as the flag for the user to define multiline search or not. Is that what you want?

Yep, the presence of this flag should be what controls whether the match is done line by line or with multiple lines.

One final thing I should mention, the doc updates you've added should also have another sentence like "This behaviour was added in ansible-core 2.12" so that people know when this can be done.

**Comment by OscarBell** [comment]
Hi @jborean93,
Are the last changes sufficient or do I need to add more stuff for this PR?

**Comment by OscarBell** [comment]
If we want to follow the design of blockinfile without (?m), we need to continue searching if there are more matches and take the last match to provide the info for the location of the anchor line. Something like
```python
if insertre.flags & re.MULTILINE:
    pos = 0
    while True:
        match = insertre.search(original, pos=pos)
        if match:
            if insertafter:
                n0 = original.decode().count('\n', 0, match.end())
            elif insertbefore:
                n0 = original.decode().count('\n', 0, match.start())
            pos = match.end()
        else:
            break
```

File: lib/ansible/modules/blockinfile.py [hunk 1]
```diff
@@ -58,6 +60,8 @@
     - If specified and no begin/ending C(marker) lines are found, the block will be inserted before the last match of specified regular expression.
     - A special value is available; C(BOF) for inserting the block at the beginning of the file.
     - If specified regular expression has no matches, the block will be inserted at the end of the file.
+    - The presence of the multiline flag (?m) in the regular expression controls whether the match is done line by line or with multiple lines.
+      This behaviour was added in ansible-core 2.12.
```
Comment (sivel) [inline lib/ansible/modules/blockinfile.py]:
```suggestion
      This behaviour was added in ansible-core 2.14.
```

**Pull request merged**